modules/renter/contractor/contracts.go

```diff
@@ -343,20 +341,38 @@ func (c *Contractor) threadedContractMaintenance() {
 	// Renew any contracts that need to be renewed.
 	c.mu.RLock()
 	var renewSet []types.FileContractID
+	refundSet := make(map[types.FileContractID]struct{})
 	for _, contract := range c.contracts {
-		if contract.GoodForRenew && c.blockHeight+c.allowance.RenewWindow >= contract.EndHeight() {
-			renewSet = append(renewSet, contract.ID)
+		if contract.GoodForRenew {
+			if c.blockHeight+c.allowance.RenewWindow >= contract.EndHeight() {
+				renewSet = append(renewSet, contract.ID)
+			} else {
+				// check if the contract has exhausted its funding and requires premature renewal.
+				c.mu.RUnlock()
+				host, _ := c.hdb.Host(contract.HostPublicKey)
+				c.mu.RLock()
+				if host.StoragePrice.Cmp(maxStoragePrice) > 0 || host.UploadBandwidthPrice.Cmp(maxUploadPrice) > 0 {
+					// skip this host if its prices are too high
+					continue
+				}
```

Comment:
It doesn't hurt to have this code here, but I believe that by the time this loop is running, another loop previously has already cycled through all of the contracts and marked the ones that no longer have acceptable prices.